CIP-0118/README-nested.md

```diff
@@ -0,0 +1,556 @@
+maintains formal guarantees, requires less off-chain communication, and does not require deposits or interaction 
+with smart contracts. 
+
+The ledger changes we describe have been developed as a result of the discussions and 
+proposals in previous CIPs, including previous versions of validation zones proposals, 
+and Transaction Swaps. The following 
+new usecases and functionality (in addition to batching) are supported by the changes :
+
+1. transactions can be batched
+- batch contains one top-level transaction, with multiple possible sub-transactions ;
+
+2. the batch must be balanced, but individual transactions in it do not have to be, so 
+- transactions in a single batch may pay each others' fees and minUTxOValue ;
+- unbalanced transactions can be interpreted as swap offers, getting resolved within a complete batch ;
+
+3. scripts data is shared across all transactions in a single batch 
+- enables script deduplication ;
+
+4. batch observer scripts enable sub-transactions to specify properties required of batches in which they may be included 
+- design inspired by script observers ;
+
+5. the top-level transaction must provide collateral scripts in its sub-transactions 
+- sub-transactions are not required to cover their own collateral ;
+
+6. a new intent we call "spend-by-output", wherein a sub-transaction may specify outputs it 
+- intends to spend, and the top-level transaction specifies the inputs that point to those outputs in the UTxO set ;
+- example of possible additional supported intents (see CPS-15)
+potentially useful for LC functionality ;
+
+Since types of supported *intents* are just specific ways in which a transaction can be underspecified,
+running scripts without supplying the `ExUnits` or collateral (as in (4) above), leaving them to be specified by 
+the top-level transaction, can be considered another type 
+of intent that is supported by this design.
+
+## Specification
+
+For the Agda specification prototype built on top of the current ledger spec, 
+see [Agda spec](https://github.com/IntersectMBO/formal-ledger-specifications/compare/polina-nested-txs).
+
+### Missing and extra assets 
+
+For a given transaction body, let `produced` and `consumed` be the produced and consumed 
+values, respectively. Then
+
+extra assets/currency `:= { aid ↦ v | aid ↦ v ∈ consumed - produced , v > 0 }`
+
+missing assets/currency `:= { aid ↦ v | aid ↦ v ∈ produced - consumed , v > 0 }`
+
+### Transaction structure changes
+
+We add the following new fields to `TxBody` : 
+
+1. `requireBatchObservers : ℙ ScriptHash`
+- set of scripts that must be run by the top-level transaction in the batch (where `TxInfo` includes info of all batch transactions)
+
+2. `spendOuts : Ix ⇀ TxOut`
+- (indexed) outputs being spent 
+
+3. `corInputs : ℙ TxIn`
+- inputs corresponding to `spentOuts`
+
+4. `subTxIds : ℙ TxId`
+-  IDs of sub-transactions in the batch
```

Comment:
Why not include `subTxs` in the `TxBody`, instead of putting them into `Tx` and having hashes here?

One less indirection that we would need to check

Reply:
The short answer is that Agda made recursive types difficult in certain cases (including ones where there is a **set** of Tx's inside of Tx). In the Haskell prototype, it is done as you suggest I believe.